File: strategy_core/telemetry.py
"""Telemetry and logging interfaces for strategy code."""

from __future__ import annotations

from typing import TYPE_CHECKING, Protocol, runtime_checkable

if TYPE_CHECKING:
    from strategy_core.models import TelemetryField, TelemetryFields


@runtime_checkable
class StrategyLogger(Protocol):
    """Minimal structured logger surface expected by strategies."""

    def debug(self, message: str, *args: object, **kwargs: object) -> None: ...

    def info(self, message: str, *args: object, **kwargs: object) -> None: ...

    def warning(self, message: str, *args: object, **kwargs: object) -> None: ...

    def error(self, message: str, *args: object, **kwargs: object) -> None: ...

    def exception(self, message: str, *args: object, **kwargs: object) -> None: ...


@runtime_checkable
class Telemetry(Protocol):
    """Unified logging and lightweight metrics surface."""

    @property
    def logger(self) -> StrategyLogger: ...

    def counter(self, name: str, value: float = 1.0, *, fields: TelemetryFields | None = None) -> None: ...

    def gauge(self, name: str, value: float, *, fields: TelemetryFields | None = None) -> None: ...

    def annotate(
        self,
        name: str,
        *,
        value: TelemetryField = None,
        fields: TelemetryFields | None = None,
    ) -> None: ...

File: native/strategy_core/src/telemetry.rs
use crate::models::{TelemetryField, TelemetryFields};

pub trait StrategyLogger {
    fn debug(&self, message: &str);
    fn info(&self, message: &str);
    fn warning(&self, message: &str);
    fn error(&self, message: &str);
    fn exception(&self, message: &str);
}

pub trait Telemetry {
    type Logger: StrategyLogger;

    fn logger(&self) -> &Self::Logger;
    fn counter(&mut self, name: &str, value: f64, fields: Option<&TelemetryFields>);
    fn gauge(&mut self, name: &str, value: f64, fields: Option<&TelemetryFields>);
    fn annotate(&mut self, name: &str, value: TelemetryField, fields: Option<&TelemetryFields>);
}
